src/ferui-components/tree-view/tree-view-component.ts

```diff
@@ -0,0 +1,399 @@
+  colorTheme: TreeViewColorTheme;
+  scrollViewArray: TreeNode<T>[];
+  hasBorders: boolean;
+  BUFFER_AMOUNT: number = 1000;
+  private rootNode: TreeNode<T>;
+  private nonRootArray: TreeNode<T>[];
+  private scrollObservable: Observable<any>;
+  private scrollSubscription: Subscription;
+  private indexChangeSubscription: Subscription;
+  private originalWidth: number;
+  private scrollPromise: boolean = false;
+  private SERVER_SIDE_COMPONENT: boolean = false;
+  private TREE_NODE_COMPONENT_HEIGHT: number = 34;
+  private DEFAULT_WIDTH: string = 'auto';
+  private DEFAULT_HEIGHT: string = '100%';
+  private DEFAULT_COLOR_THEME: TreeViewColorTheme = 'WHITE';
+  // Default Throttle time fro scroll event
+  private THROTTLE_TIME: number = 500;
+
+  constructor(private el: ElementRef) {}
+
+  /**
+   * Initializes the Tree View component and all properties needed depending on inputs configuration
+   */
+  ngOnInit(): void {
+    this.SERVER_SIDE_COMPONENT = this.dataRetriever.hasOwnProperty('getPagedChildNodeData');
+    this.treeViewStyles = {
+      width: this.config.width ? this.config.width : this.DEFAULT_WIDTH,
+      height: this.config.height ? this.config.height : this.DEFAULT_HEIGHT,
+    };
+    this.indexChangeSubscription = this.vs.vsChange.subscribe(() => this.calculateScrollerWidth());
+    this.colorTheme = this.config.colorVariation ? this.config.colorVariation : this.DEFAULT_COLOR_THEME;
+    this.border = this.hasBorders = this.config.hasBorders ? this.config.hasBorders : false;
+    if (this.SERVER_SIDE_COMPONENT) {
+      this.BUFFER_AMOUNT = this.config.bufferAmount || this.BUFFER_AMOUNT;
+      this.scrollObservable = fromEvent(this.vs.element.nativeElement, 'scroll');
+      const scrollThrottleTime = this.config.throttleTime || this.THROTTLE_TIME;
+      this.scrollSubscription = this.scrollObservable.pipe(throttleTime(scrollThrottleTime)).subscribe(() => {
+        this.handleScroll(this.vs.viewPortInfo.endIndex);
+      });
+    }
+    if (this.treeNodeData instanceof NonRootTreeNode) {
+      const emptyRootNode = this.createTreeNode(this.treeNodeData, null);
+      this.dataRetriever.getChildNodeData(emptyRootNode).then(children => {
+        this.nonRootArray = children.map(child => {
+          return this.createTreeNode(child, null);
+        });
+        this.scrollViewArray = this.nonRootArray;
+      });
+    } else {
+      this.rootNode = this.createTreeNode(this.treeNodeData, null);
+      this.scrollViewArray = [this.rootNode];
+    }
+  }
+
+  /**
+   * Destroys scroll subscription if any
+   */
+  ngOnDestroy(): void {
+    if (this.scrollSubscription) {
+      this.scrollSubscription.unsubscribe();
+    }
+    if (this.indexChangeSubscription) {
+      this.indexChangeSubscription.unsubscribe();
+    }
+  }
+
+  /**
+   * Calculate the view's children Tree Nodes widths to determine if we must increase the VS width
+   */
+  private calculateScrollerWidth(): void {
```

Comment:
You actually need to calculate the width of each nodes in FuiTreeNodeComponent directly and use a service (TreeViewUtilsService for instance) where you can update a `treeNodeMaxWidth` variable depending on your current treeNode width value. You'll be able to reset this value depending on whether i expand more children or collapse them. The node are added in the dom through a *ngFor loop in the template already. So use that loop instead.
This service will be shared between the `FuiTreeNodeComponent` and `FuiTreeViewComponent` components and you'll keep track of the maximum width value way much easier that way.
That will be much better and will avoid the use of another extra un-necessary looping rampage ;-).
Then, in this method, you'll access the service and extract the maxWidth value and you'll be able to set the width for this.vs.contentElementRef.nativeElement.style.width.

Reply:
yes but on this service you’re talking about it looks like you’re taking into consideration all FuiTreeNodeComponents which might not always be best since they’re not all on the view, we should only change the width depending on the nodes at the current view don’t you think?

Reply:
The virtual scroller will add and remove DOM elements as you scroll. So we can apply it to every `FuiTreeNodeComponent` since only the ones that are visible will be loaded in DOM.
So it will be more efficient than adding a new loop on every scroll change event... 
Don't forget that Scrolling is one of the most frequent actions the users perform, and executing JavaScript (specially loops like that) during those events makes it hard to maintain a solid 60 FPS (frames per second) scroll. The UI becomes out of sync with the user's mouse/finger/stylus. This is called scroll jank. 
That's why i think we need to let the `FuiTreeNodeComponent` update the service each time it is created (ngOnInit) thanks to virtualisation.